pkg/kubelet/apis/config/v1beta1/defaults.go

```diff
@@ -208,6 +208,9 @@ func SetDefaults_KubeletConfiguration(obj *kubeletconfigv1beta1.KubeletConfigura
 	if obj.SerializeImagePulls == nil {
 		obj.SerializeImagePulls = utilpointer.BoolPtr(true)
```

Comment:
This would override SerializeImagePulls, if a user
1. specifies `SerializeImagePulls=true` or leaves it to be the default, and
2. specifies `MaxParallelImagePulls=5`.

Having both `SerializeImagePulls=true` and `MaxParallelImagePulls=5` seems to be an invalid input, I prefer adding a validation against this case, instead of silently overriding. WDYT?

Reply:
> Having both `SerializeImagePulls=true` and `MaxParallelImagePulls=5` seems to be an invalid input, I prefer adding a validation against this case, instead of silently overriding. WDYT?

Back to what talked about the future behavior that default `SerializeImagePulls =false & MaxParallelImagePulls=5 or n`, Jordon's suggestion may help in the future case.
- defaulting `MaxParallelImagePulls=5` will override `SerializeImagePulls=true` to `SerializeImagePulls=false`. Or after the upgrade, some kubelet may fail for default value change.

The case details would be
- before the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to nil/1.
- after the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to 5. (This would fail if we add that validation.)

Reply:
> can we be smarter about defaulting so that someone doesn't have to specify _both_ `serializeImagePulls: false` and `maxParallelImagePulls: 5`?
> 
> ```
> if obj.MaxParallelImagePulls == nil || *obj.MaxParallelImagePulls < 2 {
>   obj.SerializeImagePulls = utilpointer.BoolPtr(true)
> } else {
>   obj.SerializeImagePulls = utilpointer.BoolPtr(false)
> }
> ```

It seems to be duplicate flags and we may deprecate or remove the `serializeImagePulls` once MaxParallelImagePulls is ready.

Reply:
Yeah I am thinking:
Option 1:
1. For now, keep `SerializeImagePulls=true ` and `MaxParallelImagePulls = nil` as default, and return error if `SerializeImagePulls=true` and `MaxParallelImagePulls` is set. 
2. When we are going to change the default `SerializeImagePulls` to false, we should change the defaults to `SerializeImagePulls=false ` and `MaxParallelImagePulls = N (TBD)` at the same time. 

> after the upgrade, the setting is SerializeImagePulls=true and default MaxParallelImagePulls to 5. (This would fail if we add that validation.)

This should not be the case if we are going to follow option 1.

Option 2 will be what Jordan suggests,  i.e. MaxParallelImagePulls takes higher precedence and overrides SerializeImagePulls. By doing this, when we change the default of SerializeImagePulls, we are essentially changing the default of SerializeImagePulls.

Option 3 will be SerializeImagePulls takes higher precedence, and MaxParallelImagePulls will be ignored if `SerializeImagePulls=true`.

I am leaning towards option 1, mostly to avoid silent overriding and to make things as explicit as possible.

Reply:
+1 for Option 2. 
The upgrade problem is my concern in Option 1.

Reply:
Even with option 1, we will not have any upgrade problem if we change the default of MaxParallelImagePulls and SerializeImagePull at the same time

Reply:
> The case details would be
> 
> * before the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to nil/1.
> * after the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to 5. (This would fail if we add that validation.)

For defaulting configuration, the upgrade would work as expected with both defaulting values changed. 

But if the user only set `SerializeImagePulls=true` in purpose, after the upgrade, it will be conflicted with `MaxParallelImagePulls=5`.

Reply:
Right, in that case where user explicitly sets SerializeImagePulls to true, option 2 will silently override it, which is not ideal right?

what if we go with option 1, and then when we change to default SerializeImagePulls=false, we will default MaxParallelImagePulls only if SerializeImagePulls is false? By doing this, if a user explicitly sets SerializeImagePulls=true, we will not set the MaxParallel and therefore will not break anything.

Reply:
> Right, in that case where user explicitly sets SerializeImagePulls to true, option 2 will silently override it, which is not ideal right?

Option 2 is that `MaxParallelImagePulls` is prior to `SerializeImagePulls`.
 
> what if we go with option 1, and then when we change to default SerializeImagePulls=false, we will default MaxParallelImagePulls only if SerializeImagePulls is false? By doing this, if a user explicitly sets SerializeImagePulls=true, we will not set the MaxParallel and therefore will not break anything.

Option 1 needs to set like `SerializeImagePulls` is prior to `MaxParallelImagePulls`.

Either are OK to me.

Reply:
> This would override SerializeImagePulls

sorry... my suggestion was implicitly inside a `if SerializeImagePulls == nil` block... I agree we should not override user-specified values in default.

If a user sets `serializeImagePulls: true` and `maxParallelImagePulls: 2`, that should be a validation error

Reply:
defaults are per-version, so we're only talking about the v1beta1 kubelet config file defaults, which could be independent from defaulting behavior in a future v1 config file version

I still think something like this gives the best user experience, only having to set one field:

```go
if obj.SerializeImagePulls == nil {
  if obj.MaxParallelImagePulls == nil || *obj.MaxParallelImagePulls < 2 {
    obj.SerializeImagePulls = utilpointer.BoolPtr(true)
  } else {
    obj.SerializeImagePulls = utilpointer.BoolPtr(false)
  }
}
```

Reply:
gotcha, updated the code. PTAL!